include/sparrow/layout.hpp

```diff
@@ -0,0 +1,251 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or mplied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <concepts>
+#include <cstdint>
+
+#include "sparrow/array_data.hpp"
+#include "sparrow/iterator.hpp"
+#include "sparrow/buffer.hpp"
+
+namespace sparrow
+{
+    template <class T, bool is_const>
+    class primitive_layout_iterator
+        : public iterator_base
+        <
+            primitive_layout_iterator<T, is_const>,
+            mpl::constify_t<T, is_const>,
+            std::contiguous_iterator_tag
+        >
+    {
+    public:
+
+        using self_type = primitive_layout_iterator<T, is_const>;
+        using base_type = iterator_base
+        <
+            self_type,
+            mpl::constify_t<T, is_const>,
+            std::contiguous_iterator_tag
+        >;
+        using pointer = typename base_type::pointer;
+
+        // Required so that std::ranges::end(p) is
+        // valid when p is a primitive_layout.
+        primitive_layout_iterator() = default;
```

Comment:
Yes, you might want to add a static check for the kind of iterator it is, see the concepts there: https://en.cppreference.com/w/cpp/iterator (in the "Iterator concepts" section).
Note: the check cannot be inside the type as it is not defined yet, and it cannot be made for all the types in template parameters, except if you write the check for example in the constructor, and the constructor defined outside the class definition. That's a trick for checking constraints on templates being well-formed once instantiated.

Reply:
Isn't the tag l.33 sufficient?

Reply:
The tag is for specifying the iterator type, the `static_assert` is for ensuring that the code is correct and detect whenever we break the base class.